### DISCUSSION TITLE
How to manage the TCXO in a custom board (EFR32BG21)

### DISCUSSION THREAD

**Discussion by lucaspeixotot** [Lora-net/LoRaMac-node#1082] [Q&A] [answered]
Hello, I have a custom board with the EFR32BG21 and I have a working code running with my port of this stack. I'm using the sx1262 radio. However, my board has a variant with a TCXO and I don't know exactly how to manage this crystal.

In the example of the B-L072Z-LRWAN1 board, there are these implementations in order to use the TCXO in sx1276 that probably can be used with sx1262. By the way, I would like to manage the TCXO with a pin in my MCU, not DIO3.

```
void SX1276IoTcxoInit( void )
{
    GpioInit( &TcxoPower, RADIO_TCXO_POWER, PIN_OUTPUT, PIN_PUSH_PULL, PIN_NO_PULL, 0 );
}

void SX1276SetBoardTcxo( uint8_t state )
{
    if( state == true )
    {
        if( GpioRead( &TcxoPower ) == 0 )
        { // TCXO OFF power it up.
            // Power ON the TCXO
            GpioWrite( &TcxoPower, 1 );
            DelayMs( BOARD_TCXO_WAKEUP_TIME );
        }
    }
    else
    {
        // Power OFF the TCXO
        GpioWrite( &TcxoPower, 0 );
    }
}

uint32_t SX1276GetBoardTcxoWakeupTime( void )
{
    return BOARD_TCXO_WAKEUP_TIME;
}

void SX1276Reset( void )
{
    // Enables the TCXO if available on the board design
    SX1276SetBoardTcxo( true );
    ...
}

```

So that, probably I need to call `SX1276SetBoardTcxo(false)` before my board is going to deep sleep and `SX1276SetBoardTcxo(true)` when the board wakes up, am I right? Is there something more that I need to manage? Is there some management that I need to do in my LoRa stack? Moreover, after this process (turn on/off the TCXO) I need to reinitialize the LoRa stack in order to send/recv packets?

**Comment by mluis1** [comment, marked as answer]
The TCXO management takes place at radio driver.

The easiest and recommended way to manage the TCXO using a SX126x radio is to let the radio do it automatically.
Please look at `sx1262mbxdas-board.c` file in order to see how it is done.

As you want to use a MCU pin to perform this task you can take inspiration from the B-L072Z-LRWAN1 platform and sx1276 driver.

The TCXO must be disabled after sending sleep operation mode command to the radio.
The TCXO must be enabled before sending any other operation mode command to the radio.

In order to achieve this goal the only files that require to be changed are:
- `sx126x-board.c`
- `sx126x.c`

In the future it would be nice if you could post this kind of questions on the project [Discussions](https://github.com/Lora-net/LoRaMac-node/discussions) tab. It is a better place to engage discussions and then we can agree if it is an issue or not.
If you don't mind I move this issue to the [Discussions](https://github.com/Lora-net/LoRaMac-node/discussions) tab.

**Reply by lucaspeixotot** [reply]
Thanks for your reply @mluis1. I didn't know this Discussion tab, next time I'll post here.

Regarding the post, I did understand the steps when the board is going to sleep and when it wakes up. However, is not clear to me whether after the board wakes up and I enable the TCXO again, is it needed to run some routine in the LoRaMac stack or can I use the stack as if nothing had happened?

**Reply by mluis1** [reply]
There is no change required to be done elsewhere in the code.
The only required changes are the ones on the radio driver itself.

**Reply by lucaspeixotot** [reply]
One more question, is the Lora stack letting the radio always sleeping by default if there aren't pending interrupts to resolve?

**Reply by mluis1** [reply]
Yes, by default the radio is set to sleep mode. It is only active when transmitting or receiving.
In addition if the MCU has nothing to do it is also set in low power mode.
